build/creator/ZcToken.sol

```diff
@@ -61,75 +67,89 @@ contract ZcToken is Erc20, IERC5095 {
     /// @return principalAmount The amount of principal tokens required for the withdrawal of `underlyingAmount`
-    function previewWithdraw(uint256 underlyingAmount) external override view returns (uint256 principalAmount){
+    function previewWithdraw(uint256 underlyingAmount) public override view returns (uint256 principalAmount){
         if (block.timestamp < maturity) {
             return 0;
         }
-        return (underlyingAmount * IRedeemer(redeemer).markets(protocol, underlying, maturity).maturityRate / IRedeemer(redeemer).getExchangeRate(protocol, cToken));
+
+        return (underlyingAmount * IRedeemer(redeemer).markets(protocol, underlying, maturity).maturityRate / IRedeemer(redeemer).exchangeRate(protocol, cToken));
     }
     /// @notice At or after maturity, Burns principalAmount from `owner` and sends exactly `underlyingAmount` of underlying tokens to `receiver`.
     /// @param underlyingAmount The amount of underlying tokens withdrawn
     /// @param receiver The receiver of the underlying tokens being withdrawn
-    /// @return principalAmount The amount of principal tokens burnt by the withdrawal
+    /// @return principalAmount The amount of principal tokens burnt by the withdrawal (TODO is this named return correct then?)
     function withdraw(uint256 underlyingAmount, address receiver, address holder) external override returns (uint256 principalAmount){
-        uint256 previewAmount = this.previewWithdraw(underlyingAmount);
+        // TODO removing both the `this.foo` and `external` bits of this pattern as it's simply an unnecessary misdirection. Discuss.
+        uint256 previewAmount = previewWithdraw(underlyingAmount);
         // If maturity is not yet reached
         if (block.timestamp < maturity) {
             revert Maturity(maturity);
         }
-        // Transfer logic
-        // If holder is msg.sender, skip approval check
+        // Transfer logic: If holder is msg.sender, skip approval check
         if (holder == msg.sender) {
-            redeemer.authRedeem(protocol, underlying, maturity, msg.sender, receiver, previewAmount);
-            return previewAmount;
-        }
-        else {
+            IRedeemer(redeemer).authRedeem(protocol, underlying, maturity, msg.sender, receiver, previewAmount);
+        } else {
             uint256 allowed = allowance[holder][msg.sender];
-            if (allowed >= previewAmount) {
+            if (allowed < previewAmount) {
                 revert Approvals(allowed, previewAmount);
             }
             allowance[holder][msg.sender] -= previewAmount;
-            redeemer.authRedeem(protocol, underlying, maturity, holder, receiver, previewAmount); 
-            return previewAmount;
+            IRedeemer(redeemer).authRedeem(protocol, underlying, maturity, holder, receiver, previewAmount); 
         }
+
+        // TODO why use named return statements if not using them? Discuss.
+        return previewAmount;
```

Comment:
This is confusing, especially considering we generally don't use pre-named returns. I think we should update the function signature to not have identifier on the return and leave this as is.

Reply:
i've resisted addressing all the style conflicts introduced by this contract at once. thing is a mess.

it will likely be rounded into form as the audit fixes go in.

that being said, i'll likely remove all named returns as it's not a pattern we use anywhere, and its a quick change.